src/ir/lubs.h

```diff
@@ -26,19 +26,122 @@ namespace wasm {
+  //
+  //   (ref.null any)  ;; an expression that we can update
+  //   (ref data)      ;; a type
+  //
+  // can be (ref null data), if we update that null to be
+  //
+  //   (ref.null data)
+  //
+  // It is safe to change the type of nulls, since all nulls are the same. This
+  // is a common case, as e.g. if a field has a nullable type, and it has a
+  // null assigned during creation, then that null would prevent us from using a
+  // more specific type if we do not update the null.
+  Type noteUpdatableExpression(Expression* curr) {
+    assert(!finalized);
+    // Look at the fallthrough value if there is one, but only if it has the
+    // identical type. If it has a more specific type, we may not be able to
+    // emit a LUB for it (the code still receives the original expression as an
+    // input, not the fallthrough), and if it has a less specific type then that
+    // is not helpful anyhow.
+    curr =
+      Properties::getIdenticallyTypedFallthrough(curr, passOptions, module);
+    if (auto* null = curr->dynCast<RefNull>()) {
+      updatableNulls.push_back(null);
+      updateLUBNullability();
+      return lub;
+    } else {
+      return note(curr->type);
+    }
+  }
 
   // Returns whether we noted any (reachable) value.
-  bool noted() { return lub != Type::unreachable; }
+  bool noted() { return lub != Type::unreachable || updatableNulls.size(); }
+
+  // Finalizes and returns the lub that we found. While doing so this will
+  // update the types of updatable nulls, if there were any, and if there is a
+  // useful lub to update them to.
+  Type get() {
+    if (!finalized) {
+      finalized = true;
+
+      // There are four possibilities for the lub of types and the presence of
+      // nulls:
+      //
+      //  1. We have no lub (it is unreachable) and we've seen no nulls. In this
+      //     case we've seen literally nothing, and the lub can stay as
+      //     the unreachable type.
+      //  2. We have a lub and no nulls. There is nothing more to do here: we
+      //     already know the lub.
+      //  3. We have both a lub and nulls. In this case we can update the nulls
+      //     to be whatever type we want, and we do so in order to make the most
+      //     specific type we can - which is that of the lub.
+      //  4. We have no lub, but we *do* have nulls. Set the LUB based on their
+      //     types, as we need a lub to report here, and that is a valid value.
+      //     (In theory, it could be an even more specific type, if we updated
+      //     all the nulls to the "most specific" of them, for example, but this
+      //     case does not matter much as other optimizations can handle it.)
+
+      if (updatableNulls.size()) {
+        if (lub == Type::unreachable) {
+          // This is case 4 from above: calculate the LUB from the nulls.
+          std::vector<Type> types;
+          for (auto* null : updatableNulls) {
+            types.push_back(null->type);
+          }
```

Comment:
Would it make sense to reuse `note(Type)` here rather than introducing a new vector?

Reply:
I guess that could be more efficient. It would leave noticeable traces, though, if you look at the internals, even though they should be hidden from the outside.

Reply:
Yeah, fair enough. No reason to mess with longer-lived state here.